src/tree/updater_gpu_hist.cu

```diff
@@ -623,6 +623,33 @@ __forceinline__ __device__ void CountLeft(int64_t* d_count, int val,
 template <typename GradientSumT>
 struct DeviceShard;
 
+// Instances of this type are created while creating the histogram bins for the
+// entire dataset across multiple sparse page batches. This keeps track of the number
+// of rows to process from a batch and the position from which to process on each device.
+struct RowStateOnDevice {
+  const size_t n_rows;  // Number of rows assigned to this device
+  size_t n_rows_processed;  // Number of rows processed thus far
+  size_t batch_n_rows;  // Number of rows to process from the current sparse page batch
+  size_t row_offset;  // Offset from the current sparse page batch to begin processing
+
+  explicit RowStateOnDevice(size_t nrows)
+    : n_rows(nrows), n_rows_processed(0), batch_n_rows(0), row_offset(0) {
+  }
+
+  // Advance the row state by the number of rows processed
+  void Advance() {
+    n_rows_processed += batch_n_rows;
+    CHECK_LE(n_rows_processed, n_rows);
+    batch_n_rows = row_offset = 0;
+  }
+
+  static RowStateOnDevice CreateRowStateOnDevice(size_t total_rows, size_t batch_rows) {
```

Comment:
I don't understand this static method. Why not just use a constructor?

Reply:
this is added primarily for the tests as a convenience method to create an instance of this type which can then be passed to the histogram index creation method. the type is constructed and its state initialized in 2 phases, and thus it takes only a single argument (as all the information isn't available during type construction). i added this to prevent the clutter in the tests, which may then require multiple steps to initialize and to set its state in the absence of this.

Reply:
If it's only needed in the tests, create a function in the tests. Also if you have to do this, having a second constructor overload is preferred to using a static method.

Reply:
done